File: app/src/api/kanban.ts
import axios from 'axios';
import type { TaskStatus, TaskPriority, TaskType } from '@/types';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000/api/v1';

// ─── Types ────────────────────────────────────────────────────────────────

export interface KanbanLabel {
  id: string;
  name: string;
  color: string;
  projectId?: string;
  createdAt: string;
}

export interface KanbanAssignee {
  id: string;
  email: string;
  firstName: string;
  lastName: string;
  fullName: string;
  avatarUrl?: string;
  role: string;
}

export interface KanbanTaskCard {
  id: string;
  title: string;
  description?: string;
  status: TaskStatus;
  priority: TaskPriority;
  taskType: TaskType;
  position: number;
  boardColumnId?: string;
  projectId: string;
  primaryAssigneeId?: string;
  primaryAssignee?: KanbanAssignee;
  labels: KanbanLabel[];
  dueDate?: string;
  estimatedHours?: number;
  actualHours: number;
  isOverdue: boolean;
  commentCount: number;
  assigneeCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface KanbanColumn {
  id: string;
  name: string;
  position: number;
  color: string;
  wipLimit?: number;
  columnType: string;
  tasks: KanbanTaskCard[];
  taskCount: number;
}

export interface KanbanBoardView {
  boardId: string;
  boardName: string;
  projectId: string;
  columns: KanbanColumn[];
  totalTasks: number;
  wipLimitsEnabled: boolean;
}

export interface DragMoveRequest {
  taskId: string;
  targetColumnId: string;
  newPosition: number;
  sourceColumnId?: string;
}

export interface KanbanTaskCreatePayload {
  title: string;
  project_id: string;
  board_column_id?: string;
  status?: string;
  priority?: string;
  description?: string;
  primary_assignee_id?: string;
  due_date?: string;
  estimated_hours?: number;
  position?: number;
}

// ─── Helpers ─────────────────────────────────────────────────────────────

const getHeaders = () => ({
  Authorization: `Bearer ${localStorage.getItem('token')}`,
});

function mapTask(task: any): KanbanTaskCard {
  return {
    ...task,
    boardColumnId: task.board_column_id,
    taskType: task.task_type,
    projectId: task.project_id,
    primaryAssigneeId: task.primary_assignee_id,
    dueDate: task.due_date,
    estimatedHours: task.estimated_hours,
    actualHours: task.actual_hours ?? 0,
    isOverdue: task.is_overdue ?? false,
    commentCount: task.comment_count ?? 0,
    assigneeCount: task.assignee_count ?? 0,
    createdAt: task.created_at,
    updatedAt: task.updated_at,
    labels: (task.labels ?? []).map((lbl: any) => ({
      id: lbl.id,
      name: lbl.name,
      color: lbl.color,
      projectId: lbl.project_id,
      createdAt: lbl.created_at,
    })),
    primaryAssignee: task.assignee
      ? {
          id: task.assignee.id,
          email: task.assignee.email,
          firstName: task.assignee.first_name,
          lastName: task.assignee.last_name,
          fullName: task.assignee.full_name,
          avatarUrl: task.assignee.avatar_url,
          role: task.assignee.role,
        }
      : undefined,
  };
}

// ─── API ─────────────────────────────────────────────────────────────────

export const kanbanApi = {
  /** Initialize board for a project (idempotent) */
  initBoard: async (projectId: string): Promise<{ boardId: string; created: boolean }> => {
    const res = await axios.post(`${API_URL}/kanban/boards/init/${projectId}`, null, {
      headers: getHeaders(),
    });
    return { boardId: res.data.board_id, created: res.data.created };
  },

  /** Fetch the full board view */
  getBoard: async (projectId: string): Promise<KanbanBoardView> => {
    const res = await axios.get(`${API_URL}/kanban/board/${projectId}`, {
      headers: getHeaders(),
    });
    const data = res.data;
    return {
      boardId: data.board_id,
      boardName: data.board_name,
      projectId: data.project_id,
      totalTasks: data.total_tasks,
      wipLimitsEnabled: data.wip_limits_enabled,
      columns: data.columns.map((col: any) => ({
        id: col.id,
        name: col.name,
        position: col.position,
        color: col.color,
        wipLimit: col.wip_limit,
        columnType: col.column_type,
        taskCount: col.task_count,
        tasks: col.tasks.map(mapTask),
      })),
    };
  },

  /** Create a task directly on the Kanban board */
  createTask: async (data: KanbanTaskCreatePayload): Promise<KanbanTaskCard> => {
    const res = await axios.post(`${API_URL}/kanban/tasks`, data, {
      headers: getHeaders(),
    });
    return mapTask(res.data);
  },

  /** Delete a task card */
  deleteTask: async (taskId: string): Promise<void> => {
    await axios.delete(`${API_URL}/kanban/tasks/${taskId}`, {
      headers: getHeaders(),
    });
  },

  /** Move a task via drag and drop */
  moveCard: async (moveData: DragMoveRequest): Promise<KanbanTaskCard> => {
    const res = await axios.put(
      `${API_URL}/kanban/tasks/${moveData.taskId}/move`,
      {
        task_id: moveData.taskId,
        target_column_id: moveData.targetColumnId,
        new_position: moveData.newPosition,
        source_column_id: moveData.sourceColumnId,
      },
      { headers: getHeaders() }
    );
    return mapTask(res.data);
  },

  /** Get labels for a project */
  getLabels: async (projectId: string): Promise<KanbanLabel[]> => {
    const res = await axios.get(`${API_URL}/kanban/labels/${projectId}`, {
      headers: getHeaders(),
    });
    return res.data.map((lbl: any) => ({
      id: lbl.id,
      name: lbl.name,
      color: lbl.color,
      projectId: lbl.project_id,
      createdAt: lbl.created_at,
    }));
  },
};

File: backend/app/schemas/kanban.py
"""Kanban board schemas — tasks, comments, assignments, labels, activity."""

from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field

from app.models.task import ActivityAction, TaskPriority, TaskStatus, TaskType


# ─── Label ──────────────────────────────────────────────────────────────────

class KanbanLabelCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    color: str = Field(default="#6366F1", pattern=r"^#[0-9A-Fa-f]{6}$")
    project_id: Optional[str] = None


class KanbanLabelResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    color: str
    project_id: Optional[str] = None
    created_at: datetime


# ─── Task Assignment ─────────────────────────────────────────────────────────

class AssigneeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    email: str
    first_name: str
    last_name: str
    full_name: str
    avatar_url: Optional[str] = None
    role: str


class TaskAssignmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    task_id: str
    user_id: str
    user: AssigneeResponse
    assigned_at: datetime


class TaskAssignRequest(BaseModel):
    user_id: str


# ─── Task Comment ─────────────────────────────────────────────────────────

class CommentAuthorResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    full_name: str
    avatar_url: Optional[str] = None


class TaskCommentCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=5000)
    mentions: List[str] = []
    parent_id: Optional[str] = None


class TaskCommentUpdate(BaseModel):
    content: str = Field(..., min_length=1, max_length=5000)


class TaskCommentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    task_id: str
    author_id: str
    author: CommentAuthorResponse
    content: str
    mentions: List[str]
    parent_id: Optional[str] = None
    is_edited: bool
    edited_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class PaginatedCommentsResponse(BaseModel):
    items: List[TaskCommentResponse]
    total: int
    page: int
    per_page: int
    has_more: bool


# ─── Task Activity ────────────────────────────────────────────────────────

class ActivityUserResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    full_name: str
    avatar_url: Optional[str] = None


class TaskActivityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    task_id: str
    user_id: str
    user: ActivityUserResponse
    action: ActivityAction
    description: str
    metadata: dict
    created_at: datetime


class PaginatedActivityResponse(BaseModel):
    items: List[TaskActivityResponse]
    total: int
    page: int
    per_page: int
    has_more: bool


# ─── Kanban Task Card ─────────────────────────────────────────────────────

class KanbanTaskCreate(BaseModel):
    """Create a new task card on the Kanban board."""
    title: str = Field(..., min_length=1, max_length=500)
    description: Optional[str] = None
    project_id: str
    board_column_id: Optional[str] = None
    status: TaskStatus = TaskStatus.TODO
    priority: TaskPriority = TaskPriority.MEDIUM
    task_type: TaskType = TaskType.TASK
    primary_assignee_id: Optional[str] = None
    due_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    position: float = 0.0
    label_ids: List[str] = []


class KanbanTaskUpdate(BaseModel):
    """Update a Kanban task card."""
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    priority: Optional[TaskPriority] = None
    task_type: Optional[TaskType] = None
    primary_assignee_id: Optional[str] = None
    due_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    board_column_id: Optional[str] = None
    position: Optional[float] = None


class KanbanCardMoveRequest(BaseModel):
    """Drag-and-drop card move payload."""
    task_id: str
    target_column_id: str
    new_position: float = 0.0
    source_column_id: Optional[str] = None


class KanbanTaskCardResponse(BaseModel):
    """Compact task card — returned when listing tasks on a board column."""
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    description: Optional[str] = None
    status: TaskStatus
    priority: TaskPriority
    task_type: TaskType
    position: float
    board_column_id: Optional[str] = None
    project_id: str
    primary_assignee_id: Optional[str] = None
    assignee: Optional[AssigneeResponse] = None
    labels: List[KanbanLabelResponse] = []
    due_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    actual_hours: float
    is_overdue: bool = False
    comment_count: int = 0
    assignee_count: int = 0
    created_at: datetime
    updated_at: datetime


class KanbanTaskDetailResponse(KanbanTaskCardResponse):
    """Full task detail — returned when opening a task card."""
    reporter_id: str
    reporter: Optional[AssigneeResponse] = None
    comments: List[TaskCommentResponse] = []
    assignments: List[TaskAssignmentResponse] = []
    subtask_count: int = 0
    progress_percentage: float = 0.0


class KanbanColumnWithTasksResponse(BaseModel):
    """Board column with its task cards — used for the full board view."""
    id: str
    name: str
    position: int
    color: str
    wip_limit: Optional[int] = None
    column_type: str
    tasks: List[KanbanTaskCardResponse] = []
    task_count: int = 0


class KanbanBoardViewResponse(BaseModel):
    """Full board view with all columns and their cards."""
    board_id: str
    board_name: str
    project_id: str
    columns: List[KanbanColumnWithTasksResponse]
    total_tasks: int
    wip_limits_enabled: bool = False


# ─── Bulk Reorder ─────────────────────────────────────────────────────────

class ColumnReorderRequest(BaseModel):
    """Reorder columns on a board."""
    column_ids: List[str]  # Ordered list


class TaskReorderRequest(BaseModel):
    """Reorder tasks within a column."""
    task_ids: List[str]  # Ordered list
